python/ray/serve/router.py

```diff
@@ -41,382 +43,190 @@ class Query:
     args: List[Any]
     kwargs: Dict[Any, Any]
     context: TaskContext
-
     metadata: RequestMetadata
-    async_future: Optional[asyncio.Future] = None
 
-    tick_enter_router: Optional[float] = None
+    # Fields used by backend worker to perform timing measurement.
     tick_enter_replica: Optional[float] = None
 
-    def __reduce__(self):
-        return type(self).ray_deserialize, (self.ray_serialize(), )
-
-    def ray_serialize(self):
-        # NOTE: this method is needed because Query need to be serialized and
-        # sent to the replica. However, after we send the query to the
-        # replica the async_future is still needed to retrieve the final
-        # result. Therefore we need a way to pass the information to replicas
-        # without removing async_future.
-        clone = copy.copy(self.__dict__)
-        clone.pop("async_future")
-        return pickle.dumps(clone)
 
-    @staticmethod
-    def ray_deserialize(value):
-        kwargs = pickle.loads(value)
-        return Query(**kwargs)
+class ReplicaSet:
+    """Data structure represents a set of replica actor handles"""
+
+    def __init__(self):
+        # The initialization setup before we receive them from long polling
+        # NOTE(simon): We have to do this because max_concurrent_queries
+        # and the replica handles come from different long poll keys.
+        self.max_concurrent_queries: int = 8
+        self.in_flight_queries: Dict[ActorHandle, set] = dict()
+
+        # The iterator used for load balancing among replicas.
+        # NOTE(simon): We can make this more pluggable and consider different
+        # policies like: min load, pick min of two replicas, pick replicas on
+        # the same node.
+        self.replica_iterator = itertools.cycle(self.in_flight_queries.keys())
+
+        # Used to unblock the `await all_in_flight_queries` wait. A newly added
```

Comment:
don't know what that wait is out of context, please describe it in higher level terms rather than code